File: backend/python/primeshape/geometry.py
import math
import numpy as np


def clamp(value, lower=0.0, upper=1.0):
    return float(max(lower, min(upper, value)))


def ramp(value, low, high):
    return clamp((value - low) / max(high - low, 1e-8))


def distance(a, b):
    return float(np.linalg.norm(np.asarray(a) - np.asarray(b)))


def angle(a, b, c):
    u = np.asarray(a) - np.asarray(b)
    v = np.asarray(c) - np.asarray(b)
    denominator = np.linalg.norm(u) * np.linalg.norm(v)
    if denominator < 1e-8:
        return 0.0
    return math.degrees(math.acos(clamp(float(u @ v / denominator), -1, 1)))


def bounds(points):
    points = np.asarray(points)
    low, high = points[:, :2].min(axis=0), points[:, :2].max(axis=0)
    return [float(low[0]), float(low[1]), float(high[0] - low[0]), float(high[1] - low[1])]


def serial_points(points):
    return [[round(float(v), 5) for v in p] for p in points]


def signal(code, label, score, kind="medicao", evidence=None):
    return {
        "codigo": code,
        "rotulo": label,
        "score": round(clamp(score), 3),
        "tipo": kind,
        "evidencias": evidence or [],
    }

File: frontend/js/browser/geometry.js
export const clamp = value => Math.max(0, Math.min(1, value));
export const distance = (a, b) => Math.hypot(...a.map((value, i) => value - b[i]));
export const ramp = (value, low, high) => clamp((value - low) / (high - low));
export const point = p => [p.x, p.y, p.z || 0];
export function angle(a, b, c) {
  const u = a.map((value, i) => value - b[i]);
  const v = c.map((value, i) => value - b[i]);
  return Math.acos(Math.max(-1, Math.min(1, u.reduce((sum, value, i) => sum + value * v[i], 0) / Math.max(Math.hypot(...u) * Math.hypot(...v), 1e-8)))) * 180 / Math.PI;
}
export function bounds(points) {
  const xs = points.map(p => p[0]), ys = points.map(p => p[1]);
  const x = Math.min(...xs), y = Math.min(...ys);
  return [x, y, Math.max(...xs) - x, Math.max(...ys) - y];
}
export const signal = (codigo, rotulo, score = 1, tipo = "medicao", evidencias = []) => ({ codigo, rotulo, score: clamp(score), tipo, evidencias });
